application/common/models/Reset.php

```diff
@@ -176,19 +204,64 @@ public function sendMethod()
         }
 
         if ($this->method->type == Method::TYPE_EMAIL) {
-            $this->sendEmail();
+            /*
+             * Send email to 'self' with verified email address
+             */
+            $subject = \Yii::t(
+                'app',
+                '{{idpName}} password reset request',
+                [
+                    'idpName' => \Yii::$app->params['idpName'],
+                ]
+            );
+            $this->sendEmail($this->method->value, $subject, 'self');
         } elseif ($this->method->type == Method::TYPE_PHONE) {
             $this->sendPhone();
         } else {
             throw new \Exception('Method using unknown type', 1456608781);
         }
     }
 
-    public function sendEmail()
+    /**
+     * @param string $toAddress
+     * @param string $subject
+     * @param string $view
+     * @param string|null $ccAddress
+     * @throws \Exception
+     */
+    public function sendEmail($toAddress, $subject, $view, $ccAddress = null)
     {
         /*
-         * send email to $this->method->value with reset code
+         * Generate code if needed, update attempt counter, save record, and send email
          */
+        if ($this->code === null) {
+            $this->code = Utils::getRandomDigits(\Yii::$app->params['reset']['codeLength']);
+        }
+        $this->attempts += 1;
+        if ($this->save()) {
+            $body = \Yii::$app->mailer->render(
```

Comment:
good point, I'll create an issue to revisit if we need to render different versions